File: Client/Assets/Common/Scripts/Net/Protocol/BooleanProxy.cs
#region Header
/**
 * 名称：NumberProxy
 
 * 日期：2015.11.16
 * 描述：
 **/
#endregion
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

namespace NetCore
{
    public class BooleanProxy : Proxy
    {
        #region Fields
        #endregion

        #region Properties    
        #endregion

        #region Constructors
        public BooleanProxy()
        {
    
        }
        #endregion


        #region Static Methods
    
        #endregion


        #region Private Methods
    
        #endregion

        public override bool getValue(IoBuffer ioBuffer, System.Type type, byte flag, out object value)
        {
            value =null;
            byte ft = getFlagTypes(flag);
            if (ft != Types.BOOLEAN)
            {
                Debuger.LogError("不是bool类型");
                
                return false;
            }
            byte signal = getFlagSignal(flag);
            if (signal == 0x00)
            {
                value =  false;
            }
            else if (signal == 0x01)
            {
                value =  true;
            }
            else
            {
                Debuger.LogError("未知的类型:" + signal);
                
                return false;
            }
            AddLog(value);
            return true;
            
        }

        public override bool setValue(IoBuffer ioBuffer, object value)
        {
            
		    byte flag = Types.BOOLEAN;
            if((bool)value){
                flag |= 0x01;
                ioBuffer.Write(flag);
            }
            else{
                ioBuffer.Write(flag);
            }
            AddLog(value);
            return true;		    
        }
    }
}

File: Server/Legacy/common/common-protocol/src/main/java/com/engine/common/protocol/proxy/BooleanProxy.java
package com.engine.common.protocol.proxy;

import java.io.IOException;
import java.util.concurrent.atomic.AtomicBoolean;

import org.apache.mina.core.buffer.IoBuffer;

import com.engine.common.protocol.Context;
import com.engine.common.protocol.Types;
import com.engine.common.protocol.exception.UnknowSignalException;
import com.engine.common.protocol.exception.WrongTypeException;

public class BooleanProxy extends AbstractProxy<Object> {

	@Override
	public Object getValue(Context ctx, byte flag) throws IOException {
		// ByteBuffer in = ctx.getBuffer();
		byte type = getFlagTypes(flag);
		if (type != Types.BOOLEAN) {
			throw new WrongTypeException(Types.BOOLEAN, type);
		}

		byte signal = getFlagSignal(flag);
		if (signal == 0x00) {
			return false;
		} else if (signal == 0x01) {
			return true;
		}
		throw new UnknowSignalException(type, signal);
	}

	@Override
	public void setValue(Context ctx, Object value) throws IOException {
		IoBuffer out = ctx.getBuffer();
		byte flag = Types.BOOLEAN;
		if (value instanceof Boolean) {
			Boolean bool = (Boolean) value;
			if (bool) {
				// #### 0001
				flag |= 0x01;
				out.put(flag);
			} else {
				// #### 0000
				out.put(flag);
			}
		} else if (value instanceof AtomicBoolean) {
			AtomicBoolean bool = (AtomicBoolean) value;
			if (bool.get()) {
				// #### 0001
				flag |= 0x01;
				out.put(flag);
			} else {
				// #### 0000
				out.put(flag);
			}
		}
	}

}
